xpra/server/auth/keycloak_auth.py

```diff
@@ -0,0 +1,125 @@
+
+          # Get authorization code
+          client = WebApplicationClient(KEYCLOAK_CLIENT_ID)
+          authorization_url = KEYCLOAK_SERVER_URL + 'realms/' + KEYCLOAK_REALM_NAME + '/protocol/openid-connect/auth'
+          self.salt = client.prepare_request_uri(
+                      authorization_url,
+                      redirect_uri = KEYCLOAK_REDIRECT_URI,
+                      scope = [KEYCLOAK_SCOPE],
+          )
+        else:
+          raise(NotImplementedError("Warning: only grant type \"authorization_code\" is currently supported."))
+
+    def __repr__(self):
+        return "keycloak"
+
+    def get_challenge(self, digests):
+        assert not self.challenge_sent
+        if "keycloak" not in digests:
+            log.error("Error: client does not support keycloak authentication")
+            return None
+        self.challenge_sent = True
+        return self.salt, "keycloak"
+
+    def check(self, response_json) -> bool:
+        assert self.challenge_sent
+        #log("response_json: %r", repr(response_json))
+        
+        if not response_json:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+        
+        try:
+          response = json.loads(response_json)
+        except Exception:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False 
+
+        if not isinstance(response, dict):
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+
+        log("check(%r)", repr(response))
+
+        auth_code = response.get("code")
+        error = response.get("error")
+
+        if not error and not auth_code:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+
+        if error:
+          log.error("Error: keycloak authentication failed with %s: %s", error, response.get("error_description"))
+          return False
+
+        if auth_code:
+          # Configure client
+          keycloak_openid = KeycloakOpenID(server_url=KEYCLOAK_SERVER_URL,
+                            client_id=KEYCLOAK_CLIENT_ID,
+                            realm_name=KEYCLOAK_REALM_NAME,
+                            client_secret_key=KEYCLOAK_CLIENT_SECRET_KEY)
+
+          try:
+            # Get well_known
+            config_well_know = keycloak_openid.well_know()
+            log("well_known: %r", repr(config_well_know))
+        
+            # Get token
+            token = keycloak_openid.token(code=auth_code, grant_type=[KEYCLOAK_GRANT_TYPE], redirect_uri=KEYCLOAK_REDIRECT_URI)
+
+            # Verify token
+            token_info = keycloak_openid.introspect(token['access_token'])
```

Comment:
Please use double quotes here like everywhere else.
Could `access_token` be missing from `token`? (which would cause an `KeyError`)
If so:
```
access_token = token.get("access_token")
if not access_token:
   log.error("Error: cannot authenticate, access token is missing")
   return False
```

Reply:
I don't think it could be missing, but let's do this for extra safety.